text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,144 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When going to the source code pages (as you can see already in the [compiler documentation](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877)), you will be able to click on idents to go to their definition.
+
+Once at their definition, if you click on the item name, you'll arrive on its documentation page (if it has one, otherwise nothing will happen).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It doesn't require javascript since it's only generating links, making it work on both desktop and mobile, with or without javascript enabled.
+
+So on the technical side, there isn't big issues. You can see the first part of this RFC implemented [here](https://github.com/rust-lang/rust/pull/84176).
+
+For the UI, we have the following constraints:
+
+ * Must work on mobile and desktop: we cannot rely on mouse hover events.
+ * It's better if it works without javascript (for accessibility, it's very important).
+
+With this in mind, the suggested UI would be like this:
+
+ * Links generated on idents in source code will jump to the ident definition.
+ * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
+ * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).
+
+Basically, the source code pages' UI doesn't change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?
+
+Unfortunately, on this part, the best answer I can provide is that we will need to put clear limits to what we don't want. I think we need to keep rustdoc focused on its target: documentation. We provide source code pages, so this feature allows to browse them more efficiently by using information we already have. Anything more advanced would be very likely out-of-scope.
+
+**Another concern is**: "why should we implement this in rustdoc? Aren't there already existing tools doing it?"
+
+A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?
```

Comment:
That's actually a good point... why not just make the line numbers jump to github and let it be handled there?

Reply:
That brings a lot of other constraints: we'd need to support all websites' URL to go to a file line (if it's even supported). And it doesn't even work with private repositories too.

Reply:
The other reason GitHub / GitLab links don't solve the problem is that they don't currently have code navigation supported for Rust: https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github.